Add e2e test for leaving reader comment

rachelmcr opened the pull request:
This PR fixes #477:

- It updates the existing ReaderPage class with current classes and reader behavior.
- It also adds a new reader spec to test leaving a comment from the reader.

As with the current notifications spec, the comment may not necessarily be deleted from the test site even if the notification confirms it was deleted (and the final test passes). We'll need to address that issue for both specs per https://github.com/Automattic/wp-e2e-tests/issues/535.

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+	this.timeout( mochaTimeOut );
+
+	test.before( function() {
+			driverManager.clearCookiesAndDeleteLocalStorage( driver );
```
This is indented slightly too much

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+
+	test.describe( 'Log in as commenting user', function() {
+		test.it( 'Can log in as commenting user', function() {
+			const commentingUser = 'e2eflowtestingcommenter';
```
I don't think this line is necessary

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+		test.describe( 'Leave a comment on the latest post in the Reader', function() {
+			test.it( 'Can see the Reader stream', function() {
+				this.readerPage = new ReaderPage( driver );
+				return this.readerPage.waitForPage();
```
This reads well as it's clear what it's doing, but it's not necessary per se, as creating a new instance of a page will automatically wait for that page. You can still leave this for readability if you like 😊

rachelmcr replied:
Got it! I think I'll leave this — it's a good point that it isn't necessary, but I like it for readability.

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+			test.it( 'The latest post is on the expected test site', function () {
+				const testSiteForNotifications = config.get( 'testSiteForNotifications' );
+				return this.readerPage.siteOfLatestPost().then( ( siteOfLatestPost ) => {
+					assert.equal( siteOfLatestPost, testSiteForNotifications, 'The site feed is not the expected test site' );
```
This is a great fail safe check 👍

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+
+			test.it( 'Can comment on the latest post', function() {
+				this.comment = dataHelper.randomPhrase();
+				return this.readerPage.commentOnLatestPost( this.comment );
```
It comes down to style, but it might be worth splitting out the leaving the comment from checking the comment is shown pending moderation. That way the test is clearer what it is doing.

alisterscott commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+					this.notificationsComponent = new NotificationsComponent( driver );
+					this.notificationsComponent.selectCommentByText( this.comment );
+					this.notificationsComponent.trashComment();
+					return this.notificationsComponent.waitForUndoMessage();
```
See possible fix for this in #538

hoverduck commented on specs/wp-reader-spec.js:
```diff
@@ -0,0 +1,76 @@
+			} );
+
+			test.describe( 'Delete the new comment', function() {
+				test.it( 'Can log in as test site owner', function() {
```
Can you add a `test.before()` call to `driverManager.clearCookiesAndDeleteLocalStorage( driver );` here before logging int as the other user to make sure we don't get any cross-user contamination?

rachelmcr commented:
My time with Xcode has made me forgetful about eslint, oops. Thanks for the reminder!

I've addressed most of the feedback above with my latest changes. Once #538 is merged I can add `waitForUndoMessageToDisappear()` to fix deleting the comment at the end.

alisterscott commented:
#538 is now merged if you would like to rebase and incorporate those changes here 👍

rachelmcr commented:
Ok, those changes are incorporated. Ready for another review!

rachelmcr commented:
After discussing with @hoverduck I'm going to make sure this test passes in wpcalypso and/or a local Calypso install before merging.

However, the test is currently failing in both wpcalypso and production due to this issue (the comment disappears from the reader when held in moderation): https://github.com/Automattic/wp-calypso/issues/14781

There's some discussion about how to resolve that issue, so I'm going to hold off on this test a bit longer to see how that's resolved — once resolved, I'll make any necessary adjustments to that step in the test and make sure it's passing in both environments.

rachelmcr commented:
@hoverduck @alisterscott I've added a workaround for Automattic/wp-calypso#14781 (and opened a new issue to track that), and now this test is passing. Would you mind doing another review on this?

hoverduck commented:
LGTM

alisterscott commented on lib/pages/reader-page.js:
```diff
@@ -12,24 +12,34 @@ export default class ReaderPage extends BaseContainer {
 	siteOfLatestPost() {
-		return this.driver.findElement( by.css( 'article.reader__card .site__domain' ) ).getText();
+		return this.driver.findElement( by.css( '.reader-visit-link' ) ).getAttribute( 'href' ).then( ( href ) => {
+			return href.split( '/' )[2];
```
This works and I don't think it would be fragile, but it can be a little hard to understand what it's doing at a glance.
You could try something like:
`return URL.parse( href ).hostname;`
(which requires this also `import URL from 'url';`) which is the in built URL parsing library: https://www.w3schools.com/nodejs/nodejs_url.asp
Saying that, I've probably written very similar or the same split code as you previously so I'm happy with that too

alisterscott commented on lib/pages/reader-page.js:
```diff
@@ -12,24 +12,34 @@ export default class ReaderPage extends BaseContainer {
+	}
+
+	waitForModeratedCommentToAppear() {
+		return this.driver.wait( until.elementLocated( by.css( '.comments__comment-moderation' ) ), this.explicitWaitMS, 'Could not see the comment moderation message after submitting a new comment' );
```
We have a driverHelper method which does something similar and also caters for elements which are located but not hidden on screen, see: `driverHelper.waitTillPresentAndDisplayed(`

rachelmcr commented:
Thanks for the reviews! I made those last changes to the Reader page (really appreciate that feedback, @alisterscott) and the CI check passed, so I'll merge this one.
